services-api/src/main/java/io/scalecube/services/auth/Authenticator.java

```diff
@@ -0,0 +1,20 @@
+package io.scalecube.services.auth;
+
+import reactor.core.publisher.Mono;
+
+/**
+ * Authenticator.
+ *
+ * @param <C> credentials type
+ * @param <P> principal type
+ */
+public interface Authenticator<C, P> {
```

Comment:
If you create a class for credentials please bind it like `C extends Credentials`

Reply:
Now `Authenticator` accepts `ServiceMessage` and is responsible to retrieve credentials from headers.